common/cache.go

```diff
@@ -13,90 +18,300 @@ import (
 
 type blockCacheEntry struct {
 	data []byte
-	hash []byte
 }
 
-// BlockCache contains a set of recent compact blocks in marshalled form.
+// BlockCache contains a consecutive set of recent compact blocks in marshalled form.
 type BlockCache struct {
-	MaxEntries int
+	lengthsName, blocksName string // pathnames
+	lengthsFile, blocksFile *os.File
+	starts                  []int64 // Starting offset of each block within blocksFile
+	firstBlock              int     // height of the first block in the cache (usually Sapling activation)
+	nextBlock               int     // height of the first block not in the cache
+	latestHash              []byte  // hash of the most recent (highest height) block, for detecting reorgs.
+	mutex                   sync.RWMutex
+}
+
+func (c *BlockCache) GetNextHeight() int {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.nextBlock
+}
+
+func (c *BlockCache) GetLatestHash() []byte {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.latestHash
+}
+
+//  HashMismatch indicates if the given prev-hash doesn't match the most recent block's hash
+// so reorgs can be detected.
+func (c *BlockCache) HashMismatch(prevhash []byte) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.latestHash != nil && !bytes.Equal(c.latestHash, prevhash)
+}
+
+// Start over, so that all blocks get re-downloaded (file corruption)
```

Comment:
If we ever determine that there's file corruption, it means there's a bug somewhere in the code or some really unlikely I/O error occurred, so we might want to leave a copy of the corrupted files on disk so we can investigate. That doesn't block this PR but we should open a ticket to do it at some point.

Reply:
Here's the ticket: https://github.com/zcash/lightwalletd/issues/233